cdc/cdc.cc

```diff
@@ -218,4 +219,195 @@ db_context db_context::builder::build() {
+        }
+        return log_ck;
+    }
+    partition_key stream_id(const net::inet_address& ip, unsigned int shard_id) const {
+        auto it = _streams.find(std::make_pair(ip, shard_id));
+        if (it == std::end(_streams)) {
+                throw std::runtime_error(format("No stream found for node {} and shard {}", ip, shard_id));
+        }
+        return partition_key::from_exploded(*_log_schema, { uuid_type->decompose(it->second) });
+    }
+public:
+    transformer(db_context ctx, const schema& s, const streams_type& streams)
+        : _ctx(ctx)
+        , _schema(s)
+        , _log_schema(ctx._proxy.get_db().local().find_schema(_schema.ks_name(), log_name(_schema.cf_name())))
+        , _time(utils::UUID_gen::get_time_UUID())
+        , _decomposed_time(timeuuid_type->decompose(_time))
+        , _ignore_msb_bits(ctx._proxy.get_db().local().get_config().murmur3_partitioner_ignore_msb_bits())
+        , _streams(streams)
+    { }
+
+    mutation transform(const mutation& m) const {
+        auto& t = m.token();
+        auto&& ep = _ctx._token_metadata.get_endpoint(
+                _ctx._token_metadata.first_token(t));
+        if (!ep) {
+            throw std::runtime_error(format("No owner found for key {}", m.decorated_key()));
+        }
+        auto shard_id = dht::murmur3_partitioner::shard_of(
+                t, _ctx._snitch->get_shard_count(*ep), _ignore_msb_bits);
+        mutation res(_log_schema, stream_id(ep->addr(), shard_id));
+        set_pk_columns(m.key(), 0, res);
+        return res;
+    }
+};
+
+class streams_builder {
+    const schema& _schema;
+    transformer::streams_type _streams{transformer::map_cmp{}};
+    net::inet_address _node_ip = net::inet_address();
+    unsigned int _shard_id = 0;
+    api::timestamp_type _latest_row_timestamp = api::min_timestamp;
+    utils::UUID _latest_row_stream_id = utils::UUID();
+public:
+    streams_builder(const schema& s) : _schema(s) {}
+
+    void accept_new_partition(const partition_key& key, uint32_t row_count) {
+        auto exploded = key.explode(_schema);
+        _node_ip = value_cast<net::inet_address>(inet_addr_type->deserialize(exploded[0]));
+        _shard_id = static_cast<unsigned int>(value_cast<int>(int32_type->deserialize(exploded[1])));
+        _latest_row_timestamp = api::min_timestamp;
+        _latest_row_stream_id = utils::UUID();
+    }
+
+    void accept_new_partition(uint32_t row_count) {
+        assert(false);
+    }
+
+    void accept_new_row(
+            const clustering_key& key,
+            const query::result_row_view& static_row,
+            const query::result_row_view& row) {
+        auto row_iterator = row.iterator();
+        api::timestamp_type timestamp = value_cast<db_clock::time_point>(
+                timestamp_type->deserialize(key.explode(_schema)[0])).time_since_epoch().count();
+        if (timestamp <= _latest_row_timestamp) {
+            return;
+        }
+        _latest_row_timestamp = timestamp;
```

Comment:
For just logging the partition key, this is fine. But for real data logging we'll need to split the mutation according to timestamps. Since it's super-rare for a single mutation to have multiple timestamps (I think it can be constructed with a BATCH statement) we can defer it, but we need to implement it at some point. @denesb can recommend a timestamp-based mutation splitter for you.

Reply:
Looks like I misunderstood this. Please add documentation to streams_builder to explain its role.

Reply:
I added a comment.